core/src/main/scala/com/github/mrpowers/spark/fast/tests/DatasetComparerV2.scala

```diff
@@ -0,0 +1,239 @@
+package com.github.mrpowers.spark.fast.tests
+
+import com.github.mrpowers.spark.fast.tests.DatasetUtils.DatasetOps
+import com.github.mrpowers.spark.fast.tests.DatasetComparer.maxUnequalRowsToShow
+import com.github.mrpowers.spark.fast.tests.SeqLikesExtensions.SeqExtensions
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.{Column, DataFrame, Dataset, Row}
+
+import scala.reflect.ClassTag
+
+trait DatasetComparerV2 {
+  private def countMismatchMessage(actualCount: Long, expectedCount: Long): String = {
+    s"""
+Actual DataFrame Row Count: '$actualCount'
+Expected DataFrame Row Count: '$expectedCount'
+"""
+  }
+
+  /**
+   * order ds1 column according to ds2 column order
+   */
+  def orderColumns[T](ds1: Dataset[T], ds2: Dataset[T]): Dataset[T] = {
+    ds1.select(ds2.columns.map(col).toIndexedSeq: _*).as[T](ds2.encoder)
+  }
+
+  /**
+   * Raises an error unless `actualDS` and `expectedDS` are equal. It is recommended to provide `primaryKeys` to ensure accurate and efficient
+   * comparison of rows. If primary key is not provided, will try to compare the rows based on their row number. This requires both datasets to be
+   * partitioned in the same way and become unreliable when shuffling is involved.
+   * @param primaryKeys
+   *   unique identifier for each row to ensure accurate comparison of rows. Make sure the primaryKeys are unique else the comparison will not be
+   *   accurate.
+   */
+  def assertLargeDatasetEquality[T: ClassTag](
+      actualDS: Dataset[T],
+      expectedDS: Dataset[T],
+      equals: Column = col("actual") =!= col("expected"),
+      ignoreNullable: Boolean = false,
+      ignoreColumnNames: Boolean = false,
+      ignoreColumnOrder: Boolean = false,
+      ignoreMetadata: Boolean = true,
+      primaryKeys: Seq[String] = Seq.empty,
+      truncate: Int = 500
+  ): Unit = {
+    assertLargeDatasetEquality(
+      actualDS,
+      expectedDS,
+      Right(Some(equals)),
+      ignoreNullable,
+      ignoreColumnNames,
+      ignoreColumnOrder,
+      ignoreMetadata,
+      primaryKeys,
+      truncate
+    )
+  }
+
+  /**
+   * Raises an error unless `actualDS` and `expectedDS` are equal. It is recommended to provide `primaryKeys` to ensure accurate and efficient
+   * comparison of rows. If primary key is not provided, will try to compare the rows based on their row number. This requires both datasets to be
+   * partitioned in the same way and become unreliable when shuffling is involved.
+   */
+  def assertLargeDatasetEquality[T: ClassTag](
```

Comment:
Cannot provide multiple overloads with default value so need to work around this way